keps/sig-network/1880-multiple-service-cidrs/README.md

```diff
@@ -0,0 +1,954 @@
+- 1 new controller that participates on the ServiceCIDRConfig deletion to avoid that exist IPAddresses without
+  a ServiceCIDRConfig associated. It also handles the special case for the `default` ServiceCIDRConfig.
+- 1 new controller that participates in the IPAddress deletion, that guarantees that each Service IP
+  allocated has its corresponding IPAddress object, recreating it if necessary.
+- 1 new controller that handles the bootstrap process and the default ServiceCIDRConfig.
+
+[![](https://mermaid.ink/img/pako:eNp9UstqwzAQ_BWxx5IcCqUHUwrFacGXYJLeqh620tpV0SNIcsAk_vfKjpNiJ3QvYmdGs7NIBxBOEmQgNIawUlh7NNyyVAPCtuT3StDhhPWV6yZE8kXJQvTK1jeYwD4-52RRvqFRWlFPjk17Rbel00q07G7av7c7Omm7G-HyYrXJna1UfYm5RkOzfEW5f7iGHifQxL0oX6T0FMJ_rhuqmPv6ScfE4VynbszJONxzYE_H5fL4PDaXIVkCUGsnMFLgMLn4t-DcYj23EM5GVHZwgAUY8gaVTA88LMEhfpMhDr1UUoWNjr2yS9JmJ9PoV6mi85BVqAMtAJvotq0VkEXf0Fk0_pNR1f0CyVS2dw)](https://mermaid.live/edit#pako:eNp9UstqwzAQ_BWxx5IcCqUHUwrFacGXYJLeqh620tpV0SNIcsAk_vfKjpNiJ3QvYmdGs7NIBxBOEmQgNIawUlh7NNyyVAPCtuT3StDhhPWV6yZE8kXJQvTK1jeYwD4-52RRvqFRWlFPjk17Rbel00q07G7av7c7Omm7G-HyYrXJna1UfYm5RkOzfEW5f7iGHifQxL0oX6T0FMJ_rhuqmPv6ScfE4VynbszJONxzYE_H5fL4PDaXIVkCUGsnMFLgMLn4t-DcYj23EM5GVHZwgAUY8gaVTA88LMEhfpMhDr1UUoWNjr2yS9JmJ9PoV6mi85BVqAMtAJvotq0VkEXf0Fk0_pNR1f0CyVS2dw)
+
+#### The kube-apiserver bootstrap process and the service-cidr flags
+
+Currently, the Service CIDRs are configured independently in each kube-apiserver using flags. During
+the bootstrap process, the apiserver uses the first IP of each range to create the special
+"kubernetes.default" Service. It also starts a reconcile loop, that synchronize the state of the
+bitmap used by the internal allocators with the assigned IPs to the Services.
+
+With current implementation, each kube-apiserver can boot with different ranges configured.
+There is no conflict resolution, each apiserver keep writing and deleting others apiservers allocator bitmap
+and Services.
+
+In order to be completely backwards compatible, the bootstrap process will remain the same, the
+difference is that instead of creating a bitmap based on the flags, it will create a new
+ServiceCIDRConfig object from the flags (flags configuration removal is out of scope of this KEP)
+with a special label `networking.kubernetes.io/service-ip-cidr-default` set to `"true"`.
+
+The new bootstrap process will be:
+
+```
+at startup:
+ read_flags
+ if invalid flags
+  exit
+ run default-service-ip-range controller
+ run apiserver
+
+controller:
+ watch all ServiceCIDR objects labelled from-flags
+  ignore being-deleted ranges
+ wait for first sync
+
+controller on_event:
+ if no default range matching exactly my flags
+  log
+  create a ServiceCIDR from my flags
+   generateName: "default-" or "from-flags-"
+   from-flags label: "true"
+ else if multiple
+  log
+  if multiple ranges match exactly my flags
+   sort them by creation time
+   log
+   delete all but one
```

Comment:
what happens if IPAddress objects already exist in those?

Reply:
as long as one range exists which covers the IPs, the others can be released

Reply:
yeah, this is the can of worms we have supporting multiple service-cidrs generated from flags, this is a special case that doesn't work with current implementation, when multiple apiserver have different flags the cluster just don't work ...  

let me try to figure this out

Reply:
ups, I just realized that I've commented on top of Tim's , it seems I didn't refresh before commenting and now most of my answers are late :smile: 
Tim answered, the condition is that the ranges matches the flags, so we can delete them since the IPs are covered https://github.com/kubernetes/enhancements/pull/3365#discussion_r902028320